{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6caf30d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLAST search with e-value 1e-100, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-100_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-100, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-100_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-100, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-100_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-100, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-100_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-100, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-100_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-100, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-100_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-100, word size 3, gapopen 11, and gapextend 1 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_protein_evalue1e-100_wordsize3_gapopen11_gapextend1.xml\n",
      "BLAST search with e-value 1e-50, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-50_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-50, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-50_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-50, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-50_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-50, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-50_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-50, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-50_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-50, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-50_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-50, word size 3, gapopen 11, and gapextend 1 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_protein_evalue1e-50_wordsize3_gapopen11_gapextend1.xml\n",
      "BLAST search with e-value 1e-10, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-10_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-10, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-10_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-10, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1e-10_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-10, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-10_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-10, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-10_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-10, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1e-10_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1e-10, word size 3, gapopen 11, and gapextend 1 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_protein_evalue1e-10_wordsize3_gapopen11_gapextend1.xml\n",
      "BLAST search with e-value 1, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_nucleotide_evalue1_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1, word size 11, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1_wordsize11_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1, word size 15, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1_wordsize15_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1, word size 20, gapopen 5, and gapextend 2 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_viroids_evalue1_wordsize20_gapopen5_gapextend2.xml\n",
      "BLAST search with e-value 1, word size 3, gapopen 11, and gapextend 1 has been completed. Results are saved in /home/abozar/pathogenereads/blast2/blast_protein_evalue1_wordsize3_gapopen11_gapextend1.xml\n",
      "All BLAST results have been merged into /home/abozar/pathogenereads/blast2/merged_results.xml\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline, NcbiblastxCommandline\n",
    "\n",
    "# Define the blast databases and their corresponding blast command\n",
    "blast_databases = {\n",
    "    \"nucleotide\": {\n",
    "        \"path\": \"/home/abozar/pathogenereads/Blastdatabase/Viralnuclotide/viral2024\",\n",
    "        \"command\": NcbiblastnCommandline\n",
    "    },\n",
    "    \"viroids\": {\n",
    "        \"path\": \"/home/abozar/pathogenereads/Blastdatabase/Viroids/ref_viroids_rep_genomes\",\n",
    "        \"command\": NcbiblastnCommandline\n",
    "    },\n",
    "    \"protein\": {\n",
    "        \"path\": \"/home/abozar/pathogenereads/Blastdatabase/viralprotein/viral1protein\",\n",
    "        \"command\": NcbiblastxCommandline\n",
    "    }\n",
    "}\n",
    "\n",
    "def blast_sequence(sequence_file, database_key, output_file, evalue, word_size, gapopen, gapextend):\n",
    "    # Retrieve the database path and command based on the database key\n",
    "    database_path = blast_databases[database_key][\"path\"]\n",
    "    blast_command = blast_databases[database_key][\"command\"]\n",
    "    \n",
    "    # Update the word size and gap penalties based on the database\n",
    "    if database_key == \"protein\":\n",
    "        word_size = 3  # Default word size for BLASTX\n",
    "        gapopen, gapextend = 11, 1  # Default for BLASTX with BLOSUM62\n",
    "    else:\n",
    "        # Keep the word size as is for BLASTN\n",
    "        gapopen, gapextend = 5, 2  # Default for BLASTN\n",
    "    \n",
    "    # Define the blast command with additional parameters\n",
    "    blast_cline = blast_command(\n",
    "        cmd=\"/usr/bin/blastn\" if database_key != \"protein\" else \"/usr/bin/blastx\",\n",
    "        query=sequence_file,\n",
    "        db=database_path,\n",
    "        outfmt=5,  # XML format\n",
    "        out=output_file,\n",
    "        evalue=evalue,\n",
    "        word_size=word_size,\n",
    "        gapopen=gapopen,\n",
    "        gapextend=gapextend\n",
    "    )\n",
    "    \n",
    "    # Execute the blast command\n",
    "    stdout, stderr = blast_cline()\n",
    "    if stderr:\n",
    "        print(f\"An error occurred: {stderr}\")\n",
    "    else:\n",
    "        print(f\"BLAST search with e-value {evalue}, word size {word_size}, gapopen {gapopen}, and gapextend {gapextend} has been completed. Results are saved in {output_file}\")\n",
    "\n",
    "# Set the paths\n",
    "contig_list_nucleotide = \"/home/abozar/pathogenereads/outputscontig_6051/final.contigs.fa\"\n",
    "output_directory = \"/home/abozar/pathogenereads/blast2\"\n",
    "merged_output_file = \"/home/abozar/pathogenereads/blast2/merged_results.xml\"\n",
    "\n",
    "# Check if the output directory exists, if not, create it\n",
    "if not os.path.exists(output_directory):\n",
    "    try:\n",
    "        os.makedirs(output_directory)\n",
    "    except PermissionError as e:\n",
    "        print(f\"Permission denied: {e}\")\n",
    "        exit(1)\n",
    "\n",
    "# Define the range of parameters to test\n",
    "evalue_thresholds = [1e-100, 1e-50, 1e-10, 1]\n",
    "word_sizes_nucleotide = [11, 15, 20]  # Word sizes for BLASTN\n",
    "word_sizes_protein = [3]              # Word size for BLASTX\n",
    "\n",
    "# Open the merged output file\n",
    "with open(merged_output_file, 'w') as merged_file:\n",
    "    # Loop over each combination of parameters\n",
    "    for evalue in evalue_thresholds:\n",
    "        for db_key in blast_databases:\n",
    "            # Update the word size and gap penalties based on the database\n",
    "            if db_key == \"protein\":\n",
    "                word_sizes = word_sizes_protein\n",
    "                gap_penalties = [(11, 1)]  # Default for BLASTX with BLOSUM62\n",
    "            else:\n",
    "                word_sizes = word_sizes_nucleotide\n",
    "                gap_penalties = [(5, 2)]  # Default for BLASTN\n",
    "            \n",
    "            # Loop over each word size and gap penalty\n",
    "            for word_size in word_sizes:\n",
    "                for gapopen, gapextend in gap_penalties:\n",
    "                    # Define the output file name based on the parameters and database\n",
    "                    output_file_name = f\"blast_{db_key}_evalue{evalue}_wordsize{word_size}_gapopen{gapopen}_gapextend{gapextend}.xml\"\n",
    "                    # Full path to the output file\n",
    "                    output_file = os.path.join(output_directory, output_file_name)\n",
    "                    # Perform BLAST with the current set of parameters and database\n",
    "                    blast_sequence(contig_list_nucleotide, db_key, output_file, evalue, word_size, gapopen, gapextend)\n",
    "                    # Read the content of the current output file\n",
    "                    with open(output_file, 'r') as file:\n",
    "                        content = file.read()\n",
    "                        # Write the content to the merged output file\n",
    "                        merged_file.write(content)\n",
    "\n",
    "# Print a success message\n",
    "print(f\"All BLAST results have been merged into {merged_output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f4b4567",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "def blast_xml_to_excel(xml_file, excel_file):\n",
    "    # Parse the BLAST XML output\n",
    "    with open(xml_file) as result_handle:\n",
    "        blast_records = NCBIXML.parse(result_handle)\n",
    "        \n",
    "        # Create a list to hold all the BLAST hits\n",
    "        blast_results = []\n",
    "        \n",
    "        # Iterate over each BLAST record\n",
    "        for blast_record in blast_records:\n",
    "            # Iterate over each alignment\n",
    "            for alignment in blast_record.alignments:\n",
    "                # Iterate over each hsp (high-scoring pair)\n",
    "                for hsp in alignment.hsps:\n",
    "                    # Append a dictionary with the desired information\n",
    "                    blast_results.append({\n",
    "                        'Query ID': blast_record.query_id,\n",
    "                        'Query Def': blast_record.query,\n",
    "                        'Subject ID': alignment.hit_id,\n",
    "                        'Subject Def': alignment.hit_def,\n",
    "                        'Percent Identity': hsp.identities / hsp.align_length * 100,\n",
    "                        'Alignment Length': hsp.align_length,\n",
    "                        'Mismatches': hsp.align_length - hsp.identities,\n",
    "                        'Gap Opens': hsp.gaps,\n",
    "                        'Query Start': hsp.query_start,\n",
    "                        'Query End': hsp.query_end,\n",
    "                        'Subject Start': hsp.sbjct_start,\n",
    "                        'Subject End': hsp.sbjct_end,\n",
    "                        'E-value': hsp.expect,\n",
    "                        'Bit Score': hsp.bits,\n",
    "                    })\n",
    "        \n",
    "        # Convert the list of dictionaries to a pandas DataFrame\n",
    "        df = pd.DataFrame(blast_results)\n",
    "        \n",
    "        # Write the DataFrame to an Excel file\n",
    "        df.to_excel(excel_file, index=False)\n",
    "\n",
    "# Specify the XML file path and the desired Excel file path\n",
    "xml_file_path = \"/home/abozar/pathogenereads/blast2/merged_results.xml\"\n",
    "excel_file_path = \"/home/abozar/pathogenereads/blast2/merged_results.xlsx\"\n",
    "\n",
    "# Convert the BLAST XML output to Excel format\n",
    "blast_xml_to_excel(xml_file_path, excel_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e044ae9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
